src/HeightmapDescriptor.cc

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+
+#include "ignition/rendering/HeightmapDescriptor.hh"
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+class ignition::rendering::HeightmapTexturePrivate
+{
+  /// \brief Texture size.
+  public: double size{1.0};
+
+  /// \brief Path to diffuse texture file.
+  public: std::string diffuse;
+
+  /// \brief Path to normal map file.
+  public: std::string normal;
+};
+
+//////////////////////////////////////////////////
+class ignition::rendering::HeightmapBlendPrivate
+{
+  /// \brief Minimum height to blend from.
+  public: double minHeight{0.0};
+
+  /// \brief Distance to blend.
+  public: double fadeDistance{0.0};
+};
+
+//////////////////////////////////////////////////
+class ignition::rendering::HeightmapDescriptorPrivate
+{
+  /// \brief Name used for caching
+  public: std::string name;
+
+  /// \brief Contains heightfield data.
+  public: std::shared_ptr<common::HeightmapData> data{nullptr};
+
+  /// \brief Heightmap XYZ size in meters.
+  public: math::Vector3d size{1.0, 1.0, 1.0};
+
+  /// \brief Heightmap XYZ origin in meters.
+  public: math::Vector3d position{0.0, 0.0, 0.0};
+
+  /// \brief Flag that enables/disables the terrain paging
+  public: bool useTerrainPaging{false};
+
+  /// \brief Number of samples per heightmap datum.
+  public: unsigned int sampling{2u};
```

Comment:
2 was a bad default to use actually. This does super sampling and hence doubled the number of vertices in gazebo-classic, which caused a performance hit. We should set this to 1. I believe 2 was used a long time ago when we wanted to be able to have finer control over the heightmap using the (disabled) [terrain editor](https://github.com/osrf/gazebo/tree/gazebo11/gazebo/gui/terrain).

Reply:
This is matching the [SDFormat default](https://github.com/osrf/sdformat/blob/ed198a063cfabd5be3b0d8de4b3787eea7a1f68a/sdf/1.8/heightmap_shape.sdf#L40). I'll open a PR against `sdf11` changing it there too.

Reply:
https://github.com/osrf/sdformat/pull/459